File: ios/ReadApp/Services/MangaAntiScrapingService.swift
import Foundation

struct MangaAntiScrapingProfile: Identifiable {
    let key: String
    let name: String
    let hostSuffixes: [String]
    let referer: String?
    let userAgent: String?
    let extraHeaders: [String: String]

    var id: String { key }

    func matches(host: String) -> Bool {
        let target = host.lowercased()
        for suffix in hostSuffixes {
            let s = suffix.lowercased()
            if target == s || target.hasSuffix("." + s) {
                return true
            }
        }
        return false
    }
}

final class MangaAntiScrapingService {
    static let profiles: [MangaAntiScrapingProfile] = [
        MangaAntiScrapingProfile(key: "acg456", name: "acg456", hostSuffixes: ["acg456.com", "www.acg456.com"], referer: "http://www.acg456.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "baozimh", name: "baozimh", hostSuffixes: ["baozimh.com", "www.baozimh.com", "bzcdn.net"], referer: "https://www.baozimh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "bilibili", name: "bilibili", hostSuffixes: ["manga.bilibili.com"], referer: "https://manga.bilibili.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "boodo", name: "boodo", hostSuffixes: ["boodo.qq.com"], referer: "https://boodo.qq.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "boylove", name: "boylove", hostSuffixes: ["boylove.cc"], referer: "https://boylove.cc/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "177pic", name: "177pic", hostSuffixes: ["177pic.info", "www.177pic.info"], referer: "http://www.177pic.info/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "18comic", name: "18comic", hostSuffixes: ["18comic.vip"], referer: "https://18comic.vip/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "18hmmcg", name: "18hmmcg", hostSuffixes: ["18h.mm-cg.com"], referer: "https://18h.mm-cg.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "2animx", name: "2animx", hostSuffixes: ["2animx.com", "www.2animx.com"], referer: "https://www.2animx.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "2feimh", name: "2feimh", hostSuffixes: ["2feimh.com", "www.2feimh.com"], referer: "https://www.2feimh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "3250mh", name: "3250mh", hostSuffixes: ["3250mh.com", "www.3250mh.com"], referer: "https://www.3250mh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "36mh", name: "36mh", hostSuffixes: ["36mh.com", "www.36mh.com"], referer: "https://www.36mh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "55comic", name: "55comic", hostSuffixes: ["55comic.com", "www.55comic.com"], referer: "https://www.55comic.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "77mh", name: "77mh", hostSuffixes: ["77mh.cc", "www.77mh.cc"], referer: "https://www.77mh.cc/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "copymanga", name: "copymanga", hostSuffixes: ["copymanga.tv"], referer: "https://copymanga.tv/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "dm5", name: "dm5", hostSuffixes: ["dm5.com", "www.dm5.com", "cdndm5.com"], referer: "https://www.dm5.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "dmzj", name: "dmzj", hostSuffixes: ["dmzj.com", "www.dmzj.com"], referer: "https://www.dmzj.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "gufengmh", name: "gufengmh", hostSuffixes: ["gufengmh9.com", "www.gufengmh9.com"], referer: "https://www.gufengmh9.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "iqiyi", name: "iqiyi", hostSuffixes: ["bud.iqiyi.com"], referer: "https://bud.iqiyi.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "jmzj", name: "jmzj", hostSuffixes: ["jmzj.xyz"], referer: "http://jmzj.xyz/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "kanman", name: "kanman", hostSuffixes: ["kanman.com", "www.kanman.com"], referer: "https://www.kanman.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(
            key: "kuaikan",
            name: "kuaikan",
            hostSuffixes: ["kuaikanmanhua.com", "www.kuaikanmanhua.com", "kkmh.com", "tn1.kkmh.com"],
            referer: "https://m.kuaikanmanhua.com/",
            userAgent: nil,
            extraHeaders: ["Origin": "https://m.kuaikanmanhua.com"]
        ),
        MangaAntiScrapingProfile(key: "kuimh", name: "kuimh", hostSuffixes: ["kuimh.com", "www.kuimh.com"], referer: "https://www.kuimh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "laimanhua", name: "laimanhua", hostSuffixes: ["laimanhua.net", "www.laimanhua.net"], referer: "https://www.laimanhua.net/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "manhuadb", name: "manhuadb", hostSuffixes: ["manhuadb.com", "www.manhuadb.com"], referer: "https://www.manhuadb.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "manhuafei", name: "manhuafei", hostSuffixes: ["manhuafei.com", "www.manhuafei.com"], referer: "https://www.manhuafei.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "manhuagui", name: "manhuagui", hostSuffixes: ["manhuagui.com", "www.manhuagui.com"], referer: "https://www.manhuagui.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "manhuatai", name: "manhuatai", hostSuffixes: ["manhuatai.com", "www.manhuatai.com"], referer: "https://www.manhuatai.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "manwa", name: "manwa", hostSuffixes: ["manwa.site"], referer: "https://manwa.site/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "mh1234", name: "mh1234", hostSuffixes: ["mh1234.com", "www.mh1234.com"], referer: "https://www.mh1234.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "mh160", name: "mh160", hostSuffixes: ["mh160.cc"], referer: "https://mh160.cc/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "mmkk", name: "mmkk", hostSuffixes: ["mmkk.me", "www.mmkk.me"], referer: "https://www.mmkk.me/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "myfcomic", name: "myfcomic", hostSuffixes: ["myfcomic.com", "www.myfcomic.com"], referer: "http://www.myfcomic.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "nhentai", name: "nhentai", hostSuffixes: ["nhentai.net"], referer: "https://nhentai.net/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "nsfwpicx", name: "nsfwpicx", hostSuffixes: ["picxx.icu"], referer: "http://picxx.icu/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "pufei8", name: "pufei8", hostSuffixes: ["pufei8.com", "www.pufei8.com"], referer: "http://www.pufei8.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "qiman6", name: "qiman6", hostSuffixes: ["qiman6.com", "www.qiman6.com"], referer: "http://www.qiman6.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "qimiaomh", name: "qimiaomh", hostSuffixes: ["qimiaomh.com", "www.qimiaomh.com"], referer: "https://www.qimiaomh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "qootoon", name: "qootoon", hostSuffixes: ["qootoon.net", "www.qootoon.net"], referer: "https://www.qootoon.net/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "qq", name: "qq", hostSuffixes: ["ac.qq.com"], referer: "https://ac.qq.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "sixmh6", name: "sixmh6", hostSuffixes: ["sixmh6.com", "www.sixmh6.com"], referer: "http://www.sixmh6.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "tuhao456", name: "tuhao456", hostSuffixes: ["tuhao456.com", "www.tuhao456.com"], referer: "https://www.tuhao456.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "twhentai", name: "twhentai", hostSuffixes: ["twhentai.com"], referer: "http://twhentai.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "u17", name: "u17", hostSuffixes: ["u17.com", "www.u17.com"], referer: "https://www.u17.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "webtoons", name: "webtoons", hostSuffixes: ["webtoons.com", "www.webtoons.com"], referer: "https://www.webtoons.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "wnacg", name: "wnacg", hostSuffixes: ["wnacg.org", "www.wnacg.org"], referer: "http://www.wnacg.org/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "xiuren", name: "xiuren", hostSuffixes: ["xiuren.org", "www.xiuren.org"], referer: "http://www.xiuren.org/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "ykmh", name: "ykmh", hostSuffixes: ["ykmh.com", "www.ykmh.com"], referer: "https://www.ykmh.com/", userAgent: nil, extraHeaders: [:]),
        MangaAntiScrapingProfile(key: "yymh889", name: "yymh889", hostSuffixes: ["yymh889.com"], referer: "http://yymh889.com/", userAgent: nil, extraHeaders: [:])
    ]
    static let profileKeys = profiles.map { $0.key }
    static let shared = MangaAntiScrapingService()

    private init() {}

    func resolveProfile(imageURL: URL, referer: String?) -> MangaAntiScrapingProfile? {
        guard UserPreferences.shared.isMangaAntiScrapingEnabled else { return nil }
        let enabledSites = UserPreferences.shared.mangaAntiScrapingEnabledSites
        guard !enabledSites.isEmpty else { return nil }

        let refererHost = URL(string: referer ?? "")?.host?.lowercased()
        let imageHost = imageURL.host?.lowercased()
        for profile in Self.profiles {
            if !enabledSites.contains(profile.key) { continue }
            if let refererHost, profile.matches(host: refererHost) {
                return profile
            }
            if let imageHost, profile.matches(host: imageHost) {
                return profile
            }
        }
        return nil
    }
}

File: android/src/main/java/com/readapp/data/manga/MangaAntiScrapingService.kt
package com.readapp.data.manga

import android.net.Uri

data class MangaAntiScrapingProfile(
    val key: String,
    val name: String,
    val hostSuffixes: List<String>,
    val referer: String?,
    val userAgent: String?,
    val extraHeaders: Map<String, String>
) {
    fun matches(host: String): Boolean {
        val target = host.lowercase()
        return hostSuffixes.any { suffix ->
            val s = suffix.lowercase()
            target == s || target.endsWith(".$s")
        }
    }
}

object MangaAntiScrapingService {
    val profiles = listOf(
        MangaAntiScrapingProfile("acg456", "acg456", listOf("acg456.com", "www.acg456.com"), "http://www.acg456.com/", null, emptyMap()),
        MangaAntiScrapingProfile("baozimh", "baozimh", listOf("baozimh.com", "www.baozimh.com", "bzcdn.net", "bzmh.net"), "https://www.baozimh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("bilibili", "bilibili", listOf("manga.bilibili.com"), "https://manga.bilibili.com/", null, emptyMap()),
        MangaAntiScrapingProfile("boodo", "boodo", listOf("boodo.qq.com"), "https://boodo.qq.com/", null, emptyMap()),
        MangaAntiScrapingProfile("boylove", "boylove", listOf("boylove.cc"), "https://boylove.cc/", null, emptyMap()),
        MangaAntiScrapingProfile("177pic", "177pic", listOf("177pic.info", "www.177pic.info"), "http://www.177pic.info/", null, emptyMap()),
        MangaAntiScrapingProfile("18comic", "18comic", listOf("18comic.vip"), "https://18comic.vip/", null, emptyMap()),
        MangaAntiScrapingProfile("18hmmcg", "18hmmcg", listOf("18h.mm-cg.com"), "https://18h.mm-cg.com/", null, emptyMap()),
        MangaAntiScrapingProfile("2animx", "2animx", listOf("2animx.com", "www.2animx.com"), "https://www.2animx.com/", null, emptyMap()),
        MangaAntiScrapingProfile("2feimh", "2feimh", listOf("2feimh.com", "www.2feimh.com"), "https://www.2feimh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("3250mh", "3250mh", listOf("3250mh.com", "www.3250mh.com"), "https://www.3250mh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("36mh", "36mh", listOf("36mh.com", "www.36mh.com"), "https://www.36mh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("55comic", "55comic", listOf("55comic.com", "www.55comic.com"), "https://www.55comic.com/", null, emptyMap()),
        MangaAntiScrapingProfile("77mh", "77mh", listOf("77mh.cc", "www.77mh.cc"), "https://www.77mh.cc/", null, emptyMap()),
        MangaAntiScrapingProfile("copymanga", "copymanga", listOf("copymanga.tv"), "https://copymanga.tv/", null, emptyMap()),
        MangaAntiScrapingProfile("dm5", "dm5", listOf("dm5.com", "www.dm5.com", "cdndm5.com"), "https://www.dm5.com/", null, emptyMap()),
        MangaAntiScrapingProfile("dmzj", "dmzj", listOf("dmzj.com", "www.dmzj.com"), "https://www.dmzj.com/", null, emptyMap()),
        MangaAntiScrapingProfile("gufengmh", "gufengmh", listOf("gufengmh9.com", "www.gufengmh9.com"), "https://www.gufengmh9.com/", null, emptyMap()),
        MangaAntiScrapingProfile("iqiyi", "iqiyi", listOf("bud.iqiyi.com"), "https://bud.iqiyi.com/", null, emptyMap()),
        MangaAntiScrapingProfile("jmzj", "jmzj", listOf("jmzj.xyz"), "http://jmzj.xyz/", null, emptyMap()),
        MangaAntiScrapingProfile("kanman", "kanman", listOf("kanman.com", "www.kanman.com"), "https://www.kanman.com/", null, emptyMap()),
        MangaAntiScrapingProfile(
            "kuaikan",
            "kuaikan",
            listOf("kuaikanmanhua.com", "www.kuaikanmanhua.com", "kkmh.com", "tn1.kkmh.com"),
            "https://www.kuaikanmanhua.com/",
            null,
            mapOf("Origin" to "https://www.kuaikanmanhua.com")
        ),
        MangaAntiScrapingProfile("kuimh", "kuimh", listOf("kuimh.com", "www.kuimh.com"), "https://www.kuimh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("laimanhua", "laimanhua", listOf("laimanhua.net", "www.laimanhua.net"), "https://www.laimanhua.net/", null, emptyMap()),
        MangaAntiScrapingProfile("manhuadb", "manhuadb", listOf("manhuadb.com", "www.manhuadb.com"), "https://www.manhuadb.com/", null, emptyMap()),
        MangaAntiScrapingProfile("manhuafei", "manhuafei", listOf("manhuafei.com", "www.manhuafei.com"), "https://www.manhuafei.com/", null, emptyMap()),
        MangaAntiScrapingProfile("manhuagui", "manhuagui", listOf("manhuagui.com", "www.manhuagui.com"), "https://www.manhuagui.com/", null, emptyMap()),
        MangaAntiScrapingProfile("manhuatai", "manhuatai", listOf("manhuatai.com", "www.manhuatai.com"), "https://www.manhuatai.com/", null, emptyMap()),
        MangaAntiScrapingProfile("manwa", "manwa", listOf("manwa.site"), "https://manwa.site/", null, emptyMap()),
        MangaAntiScrapingProfile("mh1234", "mh1234", listOf("mh1234.com", "www.mh1234.com"), "https://www.mh1234.com/", null, emptyMap()),
        MangaAntiScrapingProfile("mh160", "mh160", listOf("mh160.cc"), "https://mh160.cc/", null, emptyMap()),
        MangaAntiScrapingProfile("mmkk", "mmkk", listOf("mmkk.me", "www.mmkk.me"), "https://www.mmkk.me/", null, emptyMap()),
        MangaAntiScrapingProfile("myfcomic", "myfcomic", listOf("myfcomic.com", "www.myfcomic.com"), "http://www.myfcomic.com/", null, emptyMap()),
        MangaAntiScrapingProfile("nhentai", "nhentai", listOf("nhentai.net"), "https://nhentai.net/", null, emptyMap()),
        MangaAntiScrapingProfile("nsfwpicx", "nsfwpicx", listOf("picxx.icu"), "http://picxx.icu/", null, emptyMap()),
        MangaAntiScrapingProfile("pufei8", "pufei8", listOf("pufei8.com", "www.pufei8.com"), "http://www.pufei8.com/", null, emptyMap()),
        MangaAntiScrapingProfile("qiman6", "qiman6", listOf("qiman6.com", "www.qiman6.com"), "http://www.qiman6.com/", null, emptyMap()),
        MangaAntiScrapingProfile("qimiaomh", "qimiaomh", listOf("qimiaomh.com", "www.qimiaomh.com"), "https://www.qimiaomh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("qootoon", "qootoon", listOf("qootoon.net", "www.qootoon.net"), "https://www.qootoon.net/", null, emptyMap()),
        MangaAntiScrapingProfile("qq", "qq", listOf("ac.qq.com"), "https://ac.qq.com/", null, emptyMap()),
        MangaAntiScrapingProfile("sixmh6", "sixmh6", listOf("sixmh6.com", "www.sixmh6.com"), "http://www.sixmh6.com/", null, emptyMap()),
        MangaAntiScrapingProfile("tuhao456", "tuhao456", listOf("tuhao456.com", "www.tuhao456.com"), "https://www.tuhao456.com/", null, emptyMap()),
        MangaAntiScrapingProfile("twhentai", "twhentai", listOf("twhentai.com"), "http://twhentai.com/", null, emptyMap()),
        MangaAntiScrapingProfile("u17", "u17", listOf("u17.com", "www.u17.com"), "https://www.u17.com/", null, emptyMap()),
        MangaAntiScrapingProfile("webtoons", "webtoons", listOf("webtoons.com", "www.webtoons.com"), "https://www.webtoons.com/", null, emptyMap()),
        MangaAntiScrapingProfile("wnacg", "wnacg", listOf("wnacg.org", "www.wnacg.org"), "http://www.wnacg.org/", null, emptyMap()),
        MangaAntiScrapingProfile("xiuren", "xiuren", listOf("xiuren.org", "www.xiuren.org"), "http://www.xiuren.org/", null, emptyMap()),
        MangaAntiScrapingProfile("ykmh", "ykmh", listOf("ykmh.com", "www.ykmh.com"), "https://www.ykmh.com/", null, emptyMap()),
        MangaAntiScrapingProfile("yymh889", "yymh889", listOf("yymh889.com"), "http://yymh889.com/", null, emptyMap())
    )

    fun resolveProfile(imageUrl: String?, referer: String?): MangaAntiScrapingProfile? {
        val refererHost = referer?.let { Uri.parse(it).host?.lowercase() }
        val imageHost = imageUrl?.let { Uri.parse(it).host?.lowercase() }
        for (profile in profiles) {
            if (refererHost != null && profile.matches(refererHost)) return profile
            if (imageHost != null && profile.matches(imageHost)) return profile
        }
        return null
    }

    fun resolveReferer(profile: MangaAntiScrapingProfile?, referer: String?, imageUrl: String?): String? {
        val normalized = normalizeReferer(referer, imageUrl)
        if (profile?.key == "dm5" && !normalized.isNullOrBlank()) {
            return normalizeHttp(normalized)
        }
        if (profile != null) {
            return profile.referer
        }
        if (!normalized.isNullOrBlank()) return normalizeHttp(normalized)
        val host = imageUrl?.let { Uri.parse(it).host } ?: return null
        return "https://$host/"
    }

    private fun normalizeReferer(referer: String?, imageUrl: String?): String? {
        val value = referer?.trim().orEmpty()
        if (value.isEmpty()) return null
        if (value.startsWith("http://") || value.startsWith("https://")) return value
        val host = imageUrl?.let { Uri.parse(it).host } ?: return value
        val normalized = if (value.startsWith("/")) value else "/$value"
        return "https://$host$normalized"
    }

    private fun normalizeHttp(value: String): String {
        var v = value
        if (v.startsWith("http://")) v = v.replace("http://", "https://")
        if (!v.endsWith("/")) v += "/"
        return v
    }
}
